### PR TITLE
protogen(xml): improve XML importer v2

### PR DESCRIPTION
**Pull request by wenchy**
1. use `github.com/subchen/go-xmldom` instead of `github.com/antchfx/xmlquery` to parse xml content.
2. better support for `incell list` and `incell map` in xml attribute or xml text, remove support for `keyed list`.
3. support `empty struct` in both xml and yaml.
4. support `predefined` struct in xml.
5. only parse metasheet in xml protogen, and only parse content in xml confgen.
6. support `default` property in both xml and yaml.

### CODE REVIEW

File: test/functest/testdata/xml/map/Map.xml [hunk 1]
```diff
@@ -5,15 +5,53 @@
...
-    <PreEntry Name="map<string, PreEntry>" />
+    <PreEntry Name="map<string, PreEntry>" Cost="[]int32" />
     <Entry Id="map<uint32, Entry>" Value="int32" />
+
+    <Meal Id="map<string, Meal>" Time="duration|{default:"30m"}">
+        <Meat Name="map<string, Meat>" Weight="string" />
+        <Vegetables>
+            <Vegetable Name="map<string, Vegetable>|{optional:true}" Num="int32" />
+        </Vegetables>
+        <Fruits>
```
Comment (wenchy) [inline test/functest/testdata/xml/map/Map.xml:16]: Please improve message name:
`Vegetables` -> `VegetableInfo`
`Fruits` -> `FruitInfo`

File: internal/importer/xml.go [hunk 2]
```diff
@@ -54,1009 +53,504 @@ func init() {
...
-}
 
-// hasChild check if xml node has any child element node
-func hasChild(n *xmlquery.Node) bool {
-	for n := n.FirstChild; n != nil; n = n.NextSibling {
-		if n.Type == xmlquery.ElementNode {
-			return true
-		}
+	rootNode := &book.Node{}
+	parseXMLNode(doc.Root, rootNode, mode)
```
Comment (wenchy) [inline internal/importer/xml.go:150]: Must check returned error. 

Please also fix it in other postitions.

File: internal/importer/xml.go [hunk 3]
```diff
@@ -54,1009 +53,504 @@ func init() {
...
+			subNode := &book.Node{}
+			parseXMLNode(child, subNode, mode)
+			curBNode.Children = append(curBNode.Children, subNode)
 		}
-	}
-	// start with 1, e.g.: Param1, Param2, ...
-	for i, dst := 1, len(attrMap.List)-len(indexMap); ; i, dst = i+1, dst+1 {
-		index, ok := indexMap[i]
-		if !ok {
+	case Confgen:
```
Comment (wenchy) [inline internal/importer/xml.go:250]: Pleas use `default:` instead.

File: internal/importer/xml.go [hunk 4]
```diff
@@ -54,1009 +53,504 @@ func init() {
...
 		}
-		newBook.AddSheet(metaSheet)
-		newBook.AddSheet(dataSheet)
-	}
-
-	// parse meta sheet
-	if parser != nil {
-		if err := newBook.ParseMetaAndPurge(); err != nil {
-			return nil, errors.WithMessage(err, "failed to parse metasheet")
+		sheet, err := parseXMLSheet(doc, Confgen)
```
Comment (wenchy) [inline internal/importer/xml.go:97]: Please use `Default` instead.

File: internal/confgen/document_parser.go [hunk 5]
```diff
@@ -115,6 +122,9 @@ func (sp *documentParser) parseMapField(field *Field, msg protoreflect.Message,
 
 	if field.opts.Layout == tableaupb.Layout_LAYOUT_INCELL {
 		// incell map
+		if node.Kind == book.ListNode && len(node.Children) > 0 {
+			node = node.Children[0]
```
[the remark this answers is not in the document]
Comment (Kybxd) [reply internal/confgen/document_parser.go]: done

File: internal/confgen/document_parser.go [hunk 6]
```diff
@@ -128,44 +138,80 @@ func (sp *documentParser) parseMapField(field *Field, msg protoreflect.Message,
...
+						if prop.RequireUnique(field.opts.Prop) ||
+							(!prop.HasUnique(field.opts.Prop) && sp.parser.deduceMapKeyUnique(field, reflectMap)) {
+							return false, xerrors.WrapKV(xerrors.E2005(elemNode.Name), elemNode.DebugKV()...)
+						}
+					}
+					if !keyPresent && !valuePresent {
+						// key and value are both not present.
+						continue
+					}
+					reflectMap.Set(newMapKey, newMapValue)
```
[the remark this answers is not in the document]
Comment (Kybxd) [reply internal/confgen/document_parser.go]: done

File: test/functest/testdata/xml/map/Map.xml (hunk 1 above)
Comment (Kybxd) [reply test/functest/testdata/xml/map/Map.xml]: done

File: internal/importer/xml.go (hunk 2 above)
Comment (Kybxd) [reply internal/importer/xml.go]: done

File: internal/importer/xml.go (hunk 3 above)
Comment (Kybxd) [reply internal/importer/xml.go]: done

File: internal/importer/xml.go (hunk 4 above)
Comment (Kybxd) [reply internal/importer/xml.go]: done

File: internal/protogen/exporter.go [hunk 7]
```diff
@@ -295,28 +295,36 @@ func (x *sheetExporter) exportField(depth int, tagid int, field *tableaupb.Field
 			x.Imports[typeInfo.ParentFilename] = true
 		}
 	} else {
-		if field.Fields != nil {
-			// iff field is a map or list and message type is not imported.
-			nestedMsgName := prefix + "." + typeName
+		// iff field is a map or list and message type is not imported.
```
Comment (wenchy) [inline internal/protogen/exporter.go:298]: Move this comment line to below of `case field.Fields != nil:`

File: internal/confgen/document_parser.go [hunk 8]
```diff
@@ -56,6 +56,13 @@ func (sp *documentParser) parseMessage(msg protoreflect.Message, node *book.Node
 				}
 			} else {
 				fieldNode = node.FindChild(field.opts.Name)
+				if fieldNode == nil && xproto.GetFieldDefaultValue(fd) != "" {
+					// if this field has a default value, use virtual node
+					fieldNode = &book.Node{
+						Name:  node.Name,
+						Value: node.Value,
+					}
+				}
```
Comment (wenchy) [inline internal/confgen/document_parser.go:65]: This is a greet feature, please add test cases to cover this.

Comment (Kybxd) [reply internal/confgen/document_parser.go:65]: added

File: internal/protogen/exporter.go (hunk 7 above)
Comment (Kybxd) [reply internal/protogen/exporter.go]: done

**Review by Kybxd — APPROVED** [review]

**Pull request merged**